test/src/main/java/ai/timefold/solver/test/impl/score/stream/DefaultShadowVariableAwareSingleConstraintAssertion.java

```diff
@@ -0,0 +1,76 @@
+package ai.timefold.solver.test.impl.score.stream;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.impl.score.constraint.ConstraintMatchPolicy;
+import ai.timefold.solver.core.impl.score.director.stream.BavetConstraintStreamScoreDirector;
+import ai.timefold.solver.core.impl.score.stream.common.AbstractConstraintStreamScoreDirectorFactory;
+import ai.timefold.solver.test.api.score.stream.ShadowVariableAwareSingleConstraintAssertion;
+import ai.timefold.solver.test.api.score.stream.SingleConstraintAssertion;
+
+public final class DefaultShadowVariableAwareSingleConstraintAssertion<Solution_, Score_ extends Score<Score_>>
+        extends AbstractSingleConstraintAssertion<Solution_, Score_> implements ShadowVariableAwareSingleConstraintAssertion {
+
+    private final AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_, ?> scoreDirectorFactory;
+    private final Solution_ solution;
+    private boolean initialized = false;
+
+    DefaultShadowVariableAwareSingleConstraintAssertion(
+            AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_, ?> scoreDirectorFactory,
+            Solution_ solution) {
+        super(scoreDirectorFactory);
+        this.scoreDirectorFactory = scoreDirectorFactory;
+        this.solution = Objects.requireNonNull(solution);
+    }
+
+    /**
+     * The logic ensures the solution is initialized only once.
+     * This is necessary because settingAllShadowVariables can also initialize the score and constraint data.
+     * Therefore, we might miss listener events if we call the initialization steps for an already initialized solution.
+     */
+    @Override
+    void ensureInitialized() {
```

Comment:
Please find the way of not duplicating the code of this class in the other assertion class.